pkg/controller/certificaterequests/approver/approver_test.go

```diff
@@ -0,0 +1,240 @@
+	now := time.Now()
+	metaNow := metav1.NewTime(now)
+	tests := map[string]struct {
+		// key that should be passed to ProcessItem.
+		// if not set, the 'namespace/name' of the 'CertificateRequest' field will be used.
+		// if neither is set, the key will be ""
+		key string
+
+		// CertificateRequest to be synced for the test.
+		// if not set, the 'key' will be passed to ProcessItem instead.
+		request *cmapi.CertificateRequest
+
+		// expectedEvent, if set, is an 'event string' that is expected to be fired.
+		expectedEvent string
+
+		// expectedConditions is the expected set of conditions on the
+		// CertificateRequest resource if an Update is made.
+		// If nil, no update is expected.
+		// If empty, an update to the empty set/nil is expected.
+		expectedConditions []cmapi.CertificateRequestCondition
+
+		// err is the expected error text returned by the controller, if any.
+		err string
+	}{
+		"do nothing if an empty 'key' is used": {},
+		"do nothing if an invalid 'key' is used": {
+			key: "abc/def/ghi",
+		},
+		"do nothing if a key references a Certificate that does not exist": {
+			key: "namespace/name",
+		},
+		"do nothing if CertificateRequest already has 'Approved' True condition": {
+			request: &cmapi.CertificateRequest{
+				ObjectMeta: metav1.ObjectMeta{Namespace: "testns", Name: "test"},
+				Status: cmapi.CertificateRequestStatus{
+					Conditions: []cmapi.CertificateRequestCondition{
+						{
+							Type:   cmapi.CertificateRequestConditionApproved,
+							Status: cmmeta.ConditionTrue,
+						},
+					},
+				},
+			},
+		},
+		"do nothing if CertificateRequest already has 'Denied' True condition": {
+			request: &cmapi.CertificateRequest{
+				ObjectMeta: metav1.ObjectMeta{Namespace: "testns", Name: "test"},
+				Status: cmapi.CertificateRequestStatus{
+					Conditions: []cmapi.CertificateRequestCondition{
+						{
+							Type:   cmapi.CertificateRequestConditionDenied,
+							Status: cmmeta.ConditionTrue,
+						},
+					},
+				},
+			},
+		},
+		"do nothing if CertificateRequest already has 'Ready' Failed condition": {
+			request: &cmapi.CertificateRequest{
+				ObjectMeta: metav1.ObjectMeta{Namespace: "testns", Name: "test"},
+				Status: cmapi.CertificateRequestStatus{
+					Conditions: []cmapi.CertificateRequestCondition{
+						{
+							Type:   cmapi.CertificateRequestConditionReady,
+							Status: cmmeta.ConditionFalse,
+							Reason: cmapi.CertificateRequestReasonFailed,
+						},
+					},
+				},
+			},
+		},
+		"do nothing if CertificateRequest already has 'Ready' Issued condition": {
+			request: &cmapi.CertificateRequest{
+				ObjectMeta: metav1.ObjectMeta{Namespace: "testns", Name: "test"},
+				Status: cmapi.CertificateRequestStatus{
+					Conditions: []cmapi.CertificateRequestCondition{
+						{
+							Type:   cmapi.CertificateRequestConditionReady,
+							Status: cmmeta.ConditionTrue,
+							Reason: cmapi.CertificateRequestReasonIssued,
+						},
+					},
+				},
+			},
+		},
```

Comment:
I wondered if there should be an error logged if we find that the CR has been assigned any sort of Ready condition *before* the Approved or Denied conditions have been set.
The other controllers should wait check for Approved before making any attempt to sign the CR, right?

I suppose we can't control the existing External issuers, that might ignore that case, but logging an error  and creating a warning Event might be a good way of highlighting this bad behavoiur.
What do you think?

Reply:
We could do, though this would throw a bunch or error logs and events for existing CRs when the upgrade happens.

I think I am happy being permissive here for the initial release.